From 4026e853aa00b942a81752d1383067b6c9717848 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:09 -0400
Subject: [powerpc] kprobes: Fix ABIv2 issues with kprobe_lookup_name

Message-id: <1410545655-205645-441-git-send-email-dzickus@redhat.com>
Patchwork-id: 94635
O-Subject: [RHEL7 PATCH 440/626] powerpc/kprobes: Fix ABIv2 issues with kprobe_lookup_name
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 47f86b4e07afd4652ab0b092cbf493bf8b96559e
Author: Anton Blanchard <anton@samba.org>
Date:   Thu Apr 3 16:08:38 2014 +1100

    powerpc/kprobes: Fix ABIv2 issues with kprobe_lookup_name

    Use ppc_function_entry in places where we previously assumed
    function descriptors exist.

    Signed-off-by: Anton Blanchard <anton@samba.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/kprobes.h b/arch/powerpc/include/asm/kprobes.h
index 7b6feab..af15d4d 100644
--- a/arch/powerpc/include/asm/kprobes.h
+++ b/arch/powerpc/include/asm/kprobes.h
@@ -30,6 +30,7 @@
 #include <linux/ptrace.h>
 #include <linux/percpu.h>
 #include <asm/probes.h>
+#include <asm/code-patching.h>
 
 #define  __ARCH_WANT_KPROBES_INSN_SLOT
 
@@ -56,9 +57,9 @@ typedef ppc_opcode_t kprobe_opcode_t;
   if ((colon = strchr(name, ':')) != NULL) {  \
    colon++;     \
    if (*colon != '\0' && *colon != '.')  \
-    addr = *(kprobe_opcode_t **)addr; \
+    addr = (kprobe_opcode_t *)ppc_function_entry(addr); \
   } else if (name[0] != '.')    \
-   addr = *(kprobe_opcode_t **)addr;  \
+   addr = (kprobe_opcode_t *)ppc_function_entry(addr); \
  } else {       \
   char dot_name[KSYM_NAME_LEN];    \
   dot_name[0] = '.';     \
-- 
1.7.1